src/main/scala/ai/privado/languageEngine/csharp/tagger/collection/CollectionTagger.scala

```diff
@@ -0,0 +1,210 @@
+ *
+ */
+
+package ai.privado.languageEngine.csharp.tagger.collection
+
+import ai.privado.cache.RuleCache
+import ai.privado.languageEngine.java.tagger.collection.CollectionUtility
+import ai.privado.model.{Constants, RuleInfo}
+import ai.privado.tagger.PrivadoParallelCpgPass
+import io.shiftleft.codepropertygraph.generated.Cpg
+import io.shiftleft.codepropertygraph.generated.nodes.{Annotation, Method}
+import io.shiftleft.semanticcpg.language.*
+import org.slf4j.LoggerFactory
+
+import scala.collection.mutable
+
+class CollectionTagger(cpg: Cpg, ruleCache: RuleCache) extends PrivadoParallelCpgPass[RuleInfo](cpg) {
+  protected val methodUrlMap: mutable.HashMap[Long, String] = mutable.HashMap[Long, String]()
+  protected val classUrlMap: mutable.HashMap[Long, String]  = mutable.HashMap[Long, String]()
+  private val logger                                        = LoggerFactory.getLogger(this.getClass)
+  enum NodeType:
+    case Method, Class
+
+  override def generateParts(): Array[RuleInfo] =
+    ruleCache.getRule.collections.filter(_.catLevelTwo == Constants.annotations).toArray
+
+  override def runOnPart(builder: DiffGraphBuilder, ruleInfo: RuleInfo): Unit = {
+    classUrlMap.addAll(collectAnnotatedUrlsFromClasses(ruleInfo.combinedRulePattern))
+    val (methodUrls, collectionMethodsCache) = collectAnnotatedUrlsFromMethods(ruleInfo.combinedRulePattern)
+    methodUrlMap.addAll(methodUrls)
+    tagSources(builder, ruleInfo, collectionMethodsCache)
+  }
+
+  protected def tagSources(
+    builder: DiffGraphBuilder,
+    ruleInfo: RuleInfo,
+    collectionMethodsCache: List[Method]
+  ): Unit = {
+    CollectionUtility.tagDirectSources(
+      builder,
+      collectionMethodsCache,
+      ruleCache.getRule.sources,
+      ruleInfo,
+      ruleCache,
+      methodUrlMap = methodUrlMap,
+      classUrlMap = classUrlMap
+    )
+
+    CollectionUtility.tagDerivedSources(
+      cpg,
+      builder,
+      collectionMethodsCache,
+      ruleInfo,
+      ruleCache,
+      methodUrlMap = methodUrlMap,
+      classUrlMap = classUrlMap
+    )
+  }
+
+  private def getUrlFromAnnotationsCode(
+    annotation: Annotation,
+    nodeType: NodeType,
+    combinedRulePatterns: String
+  ): String = {
+    // If we have [controller] in annotation code, then replace with the class name of controller
+    //
+    //    [Route("api/some/[controller]")]                      <-- /api/some/email            [CASE A]
+    //    class EmailController {
+    //      [HttpGet("[controller]/[action]/{id}")]             <-- /api/some/email/edit/{id}  [CASE B]
+    //      public IActionResult Edit(int id)
+    //      {
+    //        return ControllerContext.MyDisplayRouteInfo(id);
+    //      }
+    //    }
+    //
+    val controllerName = if annotation.code.contains("[controller]") then {
+      nodeType match
+        case NodeType.Class =>
+          annotation.start.typeDecl
+            .where(_.name(".*Controller"))
```

Comment:
Do we need to have this check, this can prevent us from getting results for controller which don't follow this convention?

Reply:
Controllers are always named like this in C#.

Reply:
This is a requirement in ASP.NET Core as @karan-batavia said. The first part of the Class name is then used to replace `[controller]` string in attribute.